File: docs/sdks/python/wm_card_client.py
"""
WM Card OpenAPI Python 客户端示例
依赖：requests >= 2.31
安装：pip install requests
"""

import csv
import io
import time
from typing import Any

import requests


class WmCardClient:
    BASE_URL = "https://winmelon.cn/open/v1"

    def __init__(self, api_key: str, timeout: int = 30):
        if not api_key.startswith("sk_live_"):
            raise ValueError("API Key 必须以 sk_live_ 开头")
        self.api_key = api_key
        self.timeout = timeout
        self.session = requests.Session()
        self.session.headers.update(
            {
                "X-API-Key": api_key,
                "Content-Type": "application/json",
                "User-Agent": "wm-card-python/1.0",
            }
        )

    def _request(self, method: str, path: str, **kwargs) -> dict[str, Any]:
        """发请求 + 简单指数退避重试"""
        url = f"{self.BASE_URL}{path}"
        max_retries = 3
        for attempt in range(max_retries + 1):
            resp = self.session.request(method, url, timeout=self.timeout, **kwargs)
            if resp.status_code == 429 or resp.status_code >= 500:
                if attempt < max_retries:
                    time.sleep(2**attempt)
                    continue
            resp.raise_for_status()
            body = resp.json()
            if body.get("code") != "OK":
                raise RuntimeError(f"API 错误: {body}")
            return body["data"]
        raise RuntimeError("重试次数耗尽")

    # ============== 商品 ==============

    def list_products(
        self,
        keyword: str | None = None,
        status: str | None = None,
        page: int = 1,
        page_size: int = 20,
    ) -> dict:
        params = {"page": page, "pageSize": page_size}
        if keyword:
            params["keyword"] = keyword
        if status:
            params["status"] = status
        return self._request("GET", "/products", params=params)

    def get_product(self, product_id: str) -> dict:
        return self._request("GET", f"/products/{product_id}")

    def create_product(self, payload: dict) -> dict:
        return self._request("POST", "/products", json=payload)

    def update_product(self, product_id: str, payload: dict) -> dict:
        return self._request("POST", f"/products/{product_id}/update", json=payload)

    # ============== 卡密 ==============

    def list_stock(
        self,
        product_id: str,
        status: str | None = None,
        page: int = 1,
        page_size: int = 50,
    ) -> dict:
        params = {"productId": product_id, "page": page, "pageSize": page_size}
        if status:
            params["status"] = status
        return self._request("GET", "/stock", params=params)

    def import_stock(self, product_id: str, cards: list[str]) -> dict:
        """导入卡密：cards 是卡密字符串列表，自动转 CSV"""
        buf = io.StringIO()
        writer = csv.writer(buf, quoting=csv.QUOTE_MINIMAL)
        for card in cards:
            writer.writerow([card])
        return self._request(
            "POST",
            "/stock/import",
            json={"productId": product_id, "csvContent": buf.getvalue()},
        )

    def stock_stats(self, product_id: str) -> dict:
        return self._request("GET", f"/stock/stats/{product_id}")

    # ============== 订单 ==============

    def list_orders(
        self,
        page: int = 1,
        page_size: int = 20,
        status: str | None = None,
    ) -> dict:
        params = {"page": page, "pageSize": page_size}
        if status:
            params["status"] = status
        return self._request("GET", "/orders", params=params)

    def get_order(self, order_id: str) -> dict:
        return self._request("GET", f"/orders/{order_id}")


# ============== 使用示例 ==============

if __name__ == "__main__":
    client = WmCardClient(api_key="sk_live_REPLACE_ME")

    # 1. 列商品
    products = client.list_products(status="ONLINE")
    print(f"在线商品: {products['total']} 个")

    # 2. 导入卡密
    cards = ["ABC-DEF-001", "ABC-DEF-002", "ABC-DEF-003"]
    result = client.import_stock("p_01J3X...", cards)
    print(f"导入: {result['imported']} 条")

    # 3. 查订单
    orders = client.list_orders(page=1, page_size=10, status="PAID")
    for o in orders["items"]:
        print(f"订单 {o['id']}: {o.get('amount')} 元")

File: docs/sdks/node/wm-card-client.js
/**
 * WM Card OpenAPI Node.js 客户端示例
 * 依赖：Node 18+（内置 fetch）
 * 无需额外 npm install
 */

const BASE_URL = "https://winmelon.cn/open/v1";

export class WmCardClient {
  /**
   * @param {string} apiKey  sk_live_ 开头的商户 API Key
   * @param {{ timeout?: number }} [options]
   */
  constructor(apiKey, options = {}) {
    if (!apiKey || !apiKey.startsWith("sk_live_")) {
      throw new Error("API Key 必须以 sk_live_ 开头");
    }
    this.apiKey = apiKey;
    this.timeout = options.timeout ?? 30_000;
  }

  /**
   * @param {string} method
   * @param {string} path
   * @param {{ query?: Record<string, any>, body?: any }} [opts]
   */
  async _request(method, path, opts = {}) {
    const url = new URL(`${BASE_URL}${path}`);
    if (opts.query) {
      for (const [k, v] of Object.entries(opts.query)) {
        if (v !== undefined && v !== null) url.searchParams.set(k, String(v));
      }
    }
    const maxRetries = 3;
    let lastErr;
    for (let attempt = 0; attempt <= maxRetries; attempt++) {
      const controller = new AbortController();
      const timer = setTimeout(() => controller.abort(), this.timeout);
      try {
        const resp = await fetch(url, {
          method,
          headers: {
            "X-API-Key": this.apiKey,
            "Content-Type": "application/json",
            "User-Agent": "wm-card-node/1.0",
          },
          body: opts.body ? JSON.stringify(opts.body) : undefined,
          signal: controller.signal,
        });
        clearTimeout(timer);
        if (resp.status === 429 || resp.status >= 500) {
          if (attempt < maxRetries) {
            await sleep(2 ** attempt * 1000);
            continue;
          }
        }
        const json = await resp.json();
        if (json.code !== "OK") {
          throw new Error(`API 错误 [${resp.status}]: ${json.message} (requestId=${json.requestId})`);
        }
        return json.data;
      } catch (err) {
        clearTimeout(timer);
        lastErr = err;
        if (err.name === "AbortError" && attempt < maxRetries) {
          await sleep(2 ** attempt * 1000);
          continue;
        }
        throw err;
      }
    }
    throw lastErr ?? new Error("重试次数耗尽");
  }

  // ============== 商品 ==============

  /** @param {{ keyword?: string, status?: 'ONLINE'|'OFFLINE'|'SOLD_OUT', page?: number, pageSize?: number }} q */
  listProducts(q = {}) {
    return this._request("GET", "/products", { query: { page: 1, pageSize: 20, ...q } });
  }

  /** @param {string} id */
  getProduct(id) {
    return this._request("GET", `/products/${id}`);
  }

  /** @param {Record<string, any>} payload */
  createProduct(payload) {
    return this._request("POST", "/products", { body: payload });
  }

  /** @param {string} id @param {Record<string, any>} payload */
  updateProduct(id, payload) {
    return this._request("POST", `/products/${id}/update`, { body: payload });
  }

  // ============== 卡密 ==============

  /** @param {string} productId @param {{ status?: string, page?: number, pageSize?: number }} [q] */
  listStock(productId, q = {}) {
    return this._request("GET", "/stock", { query: { productId, page: 1, pageSize: 50, ...q } });
  }

  /**
   * 导入卡密
   * @param {string} productId
   * @param {string[]} cards 卡密字符串数组
   */
  importStock(productId, cards) {
    const csvContent = cards.map((c) => escapeCsv(c)).join("\n");
    return this._request("POST", "/stock/import", {
      body: { productId, csvContent },
    });
  }

  /** @param {string} productId */
  stockStats(productId) {
    return this._request("GET", `/stock/stats/${productId}`);
  }

  // ============== 订单 ==============

  /** @param {{ page?: number, pageSize?: number, status?: string }} [q] */
  listOrders(q = {}) {
    return this._request("GET", "/orders", { query: { page: 1, pageSize: 20, ...q } });
  }

  /** @param {string} id */
  getOrder(id) {
    return this._request("GET", `/orders/${id}`);
  }
}

function escapeCsv(value) {
  if (/[",\n\r]/.test(value)) {
    return `"${value.replace(/"/g, '""')}"`;
  }
  return value;
}

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

// ============== 使用示例 ==============

// const client = new WmCardClient("sk_live_REPLACE_ME");
//
// // 1. 列商品
// const products = await client.listProducts({ status: "ONLINE" });
// console.log(`在线商品: ${products.total} 个`);
//
// // 2. 导入卡密
// const result = await client.importStock("p_01J3X...", [
//   "ABC-DEF-001",
//   "ABC-DEF-002",
//   "ABC-DEF-003",
// ]);
// console.log(`导入: ${result.imported} 条`);
//
// // 3. 查订单
// const orders = await client.listOrders({ page: 1, pageSize: 10, status: "PAID" });
// for (const o of orders.items) {
//   console.log(`订单 ${o.id}: ${o.amount} 元`);
// }
